[SeveralApps]  Improving and adding missing python scripts for adaptive remeshing

loumalouomega opened the pull request:
This adds some missing scripts fro adaptive remeshing. Specially the analysis for FluidDynamicsApplication 

(This is in order to make the examples on the Examples repository to  work and specially in order to make easier to correct/update in the future)

loumalouomega committed:
Missing main for StructuralMechanicsApplication and ContactStructuralMechanicsApplications

loumalouomega committed:
Minor

loumalouomega committed:
FluidDynamicsApplication adaptive remshinf analysis script

loumalouomega committed:
Adding extra debug info

philbucher commented:
I have a couple of comments related to the imports
I would push directly to this branch (tomorrow) if this is ok for you

loumalouomega commented:
But run the examples in the examples repository. If your concern is that
the MeshingApplication is imported I can tell you that without it cannot
find the MMG_process.py

El dom., 31 mar. 2019 19:10, Philipp Bucher <notifications@github.com>
escribió:

> I have a couple of comments related to the imports
> I would push directly to this branch (tomorrow) if this is ok for you
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/4558#issuecomment-478359746>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATEMgLPOF8PtfviyamCtoOxMP6nEaFq2ks5vcOvngaJpZM4cUaEU>
> .
>

philbucher commented:
Ok I will check

philbucher committed:
correcting main-structural

philbucher committed:
clean adaptive structural

philbucher committed:
clean structural adaptive utils

philbucher committed:
clean structural adaptive srcritps

philbucher committed:
clean adaptive fluid analysis

philbucher committed:
clean adaptive contact scripts

philbucher committed:
correcting main-contact

philbucher committed:
correcting adaptive contact

philbucher commented:
I cleaned the imports, there were many unused ones

I also checked the examples and pushed some corrections, now they should work also without importing the MeshingApp

I don't have mmg yet, could you please test the cases you are concerned about?
The tests in Kratos are running

Note: `import python_script` will stop working at some point, the safe way is to add the application-name like this: `from kratos.AppName.python_script as py_script`
=> this also imports the application

loumalouomega commented:
> I cleaned the imports, there were many unused ones
> 
> I also checked the examples and pushed some corrections, now they should work also without importing the MeshingApp
> 
> I don't have mmg yet, could you please test the cases you are concerned about?
> The tests in Kratos are running
> 
> Note: `import python_script` will stop working at some point, the safe way is to add the application-name like this: `from kratos.AppName.python_script as py_script`
> => this also imports the application

I will

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega commented:
> > I cleaned the imports, there were many unused ones
> > I also checked the examples and pushed some corrections, now they should work also without importing the MeshingApp
> > I don't have mmg yet, could you please test the cases you are concerned about?
> > The tests in Kratos are running
> > Note: `import python_script` will stop working at some point, the safe way is to add the application-name like this: `from kratos.AppName.python_script as py_script`
> > => this also imports the application
> 
> I will

It works fine

rubenzorrilla commented on applications/FluidDynamicsApplication/python_scripts/adaptative_remeshing_fluid_dynamics_analysis.py:
```diff
@@ -0,0 +1,104 @@
+
+        while self.time < self.end_time:
+            self.time = self._GetSolver().AdvanceInTime(self.time)
+            # We reinitialize if remeshed previously
```
Shouldn't the re-meshing be done always in the same moment? These if statements can be easily misunderstood...

loumalouomega replied:
Nope, you may be interested in doing the remeshing during the simulation, not before, and then you need to reinitialize the whole simulation

rubenzorrilla replied:
But this means that you might do the re-meshing before and after the `InitializeSolutionStep()`. When does this have sense?
If the re-meshing is to be done once, and we want to call it as a process, I'd rather go for a interval-type solution as we use in the processes. Then you won't probably need to check this kind of casuistic.

rubenzorrilla reviewed (commented):
Why it is needed to have a different analysis stage for remeshing?

loumalouomega commented:
> Why it is needed to have a different analysis stage for remeshing?

Because if you remesh you need to reinitialize everything, maybe we can integrate the remeshing into the base analysis stage (if ModelPart is modified then reinitilize the whole system)

philbucher commented:
> > Why it is needed to have a different analysis stage for remeshing?
> 
> Because if you remesh you need to reinitialize everything, maybe we can integrate the remeshing into the base analysis stage (if ModelPart is modified then reinitilize the whole system)

hm I am not sure abt this ...
This is quite specific to problems with a mesh
What abt DEM, MPM etc?

Also the current solution is too messy imo to go mainstream

rubenzorrilla commented:
> > Why it is needed to have a different analysis stage for remeshing?
> 
> Because if you remesh you need to reinitialize everything, maybe we can integrate the remeshing into the base analysis stage (if ModelPart is modified then reinitilize the whole system)

This is my point, I'd go for a common (as common as possible as @philbucher says) rather than adding new analysis stage files to all the apps.

loumalouomega commented:
> > > Why it is needed to have a different analysis stage for remeshing?
> > 
> > 
> > Because if you remesh you need to reinitialize everything, maybe we can integrate the remeshing into the base analysis stage (if ModelPart is modified then reinitilize the whole system)
> 
> This is my point, I'd go for a common (as common as possible as @philbucher says) rather than adding new analysis stage files to all the apps.

Agree, but modifying the base analysis requires an agreement of many people, which may be not possible. I can in this PR unify the derived analysis, and we can have a discussion about to include it on the base class

philbucher commented:
before modifying the `AnalysisStage` I would recommend to first collect the necessary operations in a common utility-file in the core, this way you reduce the duplication significantly

As I said before, I am not sure if the remeshing stuff can go into the base-class, since it is not applicable for all simulations

loumalouomega commented:
> before modifying the `AnalysisStage` I would recommend to first collect the necessary operations in a common utility-file in the core, this way you reduce the duplication significantly
> 
> As I said before, I am not sure if the remeshing stuff can go into the base-class, since it is not applicable for all simulations

Yes, that was what I meant

loumalouomega committed:
Creating auxiliar utilities

loumalouomega commented:
I updated for now only the Fluid, if you agree I will do the same for the other analysis

loumalouomega commented on applications/FluidDynamicsApplication/python_scripts/fluid_dynamics_analysis.py:
```diff
@@ -34,6 +34,14 @@ def __init__(self,model,parameters):
+        It can be overridden by derived classes
+        """
+
+        import auxiliar_methods_adaptative_remeshing
```
```suggestion
        import KratosMultiphysics.auxiliar_methods_adaptative_remeshing as auxiliar_methods_adaptative_remeshing
```

loumalouomega committed:
import KratosMultiphysics.auxiliar_methods_adaptative_remeshing as auxiliar_methods_adaptative_remeshing

loumalouomega commented:
> dexription is missing
 There are desciptions

philbucher commented on kratos/python_scripts/auxiliar_methods_adaptative_remeshing.py:
```diff
@@ -0,0 +1,44 @@
+# Importing Kratos
+import KratosMultiphysics
+
+def AdaptativeRemeshingRunSolutionLoop(analysis):
```
it looks better now but still it is quite messy IMO

not blocking, but we should carefully evaluate this, otherwise in the end I am afraid none will understand any more what is going on

loumalouomega replied:
OK

philbucher commented on kratos/python_scripts/auxiliar_methods_adaptative_remeshing.py:
```diff
@@ -0,0 +1,44 @@
+    while analysis.time < analysis.end_time:
+        analysis.time = analysis._GetSolver().AdvanceInTime(analysis.time)
+        # We reinitialize if remeshed previously
+        if root_model_part.Is(KratosMultiphysics.MODIFIED):
```
If we add this mayor functionalitiy here then I think it would make sense to add a `REMESHED` flag to indicate exactly what is going on

loumalouomega replied:
We cannot adds many more flags...

philbucher replied:
Thats why it should be discussed
For a major functionality I think it should ate least be considered

loumalouomega replied:
Ok, but take into account that many implementations (not just python, but cpp) assume that flag, and should be changed

loumalouomega committed:
More description

loumalouomega commented:
> please use the new method from the baseclass

What do you mean?

philbucher commented on kratos/python_scripts/auxiliar_methods_adaptative_remeshing.py:
```diff
@@ -0,0 +1,50 @@
+    computing_model_part = analysis._GetSolver().GetComputingModelPart()
+    root_model_part = computing_model_part.GetRootModelPart()
+
+    while analysis.time < analysis.end_time:
```
```suggestion
    while self.KeepAdvancingSolutionLoop():
```

loumalouomega replied:
What is this?, is this method implemented?

loumalouomega replied:
In fact it is analysis.KeepAdvancingSolutionLoop():

loumalouomega committed:
Suggestion

rubenzorrilla commented on applications/FluidDynamicsApplication/python_scripts/adaptative_remeshing_fluid_dynamics_analysis.py:
```diff
@@ -0,0 +1,104 @@
+        list_of_processes = super(AdaptativeRemeshingFluidDynamicsAnalysis, self)._CreateProcesses(parameter_name, initialization_order)
+
+        if parameter_name == "processes":
+            processes_block_names = ["recursive_remeshing_process"]
```
Why is it always recursive? I'd call it `mesh_adaptivity_procresses` for the sake of generality.

loumalouomega commented:
If you remesh, no matter what, you need to restart all, because some
strategies assume sizes on rhs and lhs and may have issues. (except if you
do after reading and just before the initialize)

El lun., 1 abr. 2019 22:04, Rubén Zorrilla <notifications@github.com>
escribió:

> *@rubenzorrilla* commented on this pull request.
> ------------------------------
>
> In
> applications/FluidDynamicsApplication/python_scripts/adaptative_remeshing_fluid_dynamics_analysis.py
> <https://github.com/KratosMultiphysics/Kratos/pull/4558#discussion_r271028783>
> :
>
> > +
> +    It can be imported and used as "black-box"
> +    """
> +
> +    def RunSolutionLoop(self):
> +        """This function executes the solution loop of the AnalysisStage
> +        It can be overridden by derived classes
> +        """
> +
> +        # If we remesh using a process
> +        computing_model_part = self._GetSolver().GetComputingModelPart()
> +        root_model_part = computing_model_part.GetRootModelPart()
> +
> +        while self.time < self.end_time:
> +            self.time = self._GetSolver().AdvanceInTime(self.time)
> +            # We reinitialize if remeshed previously
>
> But this means that you might do the re-meshing before and after the
> InitializeSolutionStep(). When does this have sense?
> If the re-meshing is to be done once, and we want to call it as a process,
> I'd rather go for a interval-type solution as we use in the processes.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/4558#discussion_r271028783>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATEMgObM5lqYGnRcBSSfab--O29bWNL3ks5vcmY-gaJpZM4cUaEU>
> .
>

rubenzorrilla commented:
> …

I know this... What I'm asking is in which situation you might want to remesh before and after the InitializeSolutionStep()?

loumalouomega commented:
> > …
> 
> I know this... What I'm asking is in which situation you might want to remesh before and after the InitializeSolutionStep()?

If you want to remesh during simulation the proper thing to do is do it after the initialize, man y variables, BC's and so on are assigned during the InitializeSolutionSTep. If you want to preserve them you need to do it just after

loumalouomega committed:
Replacing by mesh_adaptivity_processes

loumalouomega committed:
Clean up includes

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Codacy suggestions

loumalouomega commented:
The comments have been addressed @philbucher @rubenzorrilla

philbucher commented:
Our comments yes, but you are missing the approval from @KratosMultiphysics/technical-committee since this PR significantly changes important functionalities

loumalouomega commented:
> Our comments yes, but you are missing the approval from @KratosMultiphysics/technical-committee since this PR significantly changes important functionalities

OK

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Missing parameter

loumalouomega committed:
Some refactoring to reduce code duplication

adityaghantasala commented on applications/StructuralMechanicsApplication/python_scripts/adaptative_remeshing_structural_mechanics_static_solver.py:
```diff
@@ -3,21 +3,11 @@
 # Importing the Kratos Library
 import KratosMultiphysics
```
Should this really be a solver ? 
Cannot it be done via a process ?

loumalouomega replied:
Nope, it replaces functions from the base solver

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

GuillermoCasas commented:
In my opinion this capability should be implemented in the base analysis/solver. I just cannot see how our abstract classes are general enough if they cannot include remeshing naturally.  

However, I this way of extending the capabilities of the base classes (through a function wrapper) can work as a way to test them before they can be added to the base classes.

loumalouomega commented:
> In my opinion this capability should be implemented in the base analysis/solver. I just cannot see how our abstract classes are general enough if they cannot include remeshing naturally.
> 
> However, I this way of extending the capabilities of the base classes (through a function wrapper) can work as a way to test them before they can be added to the base classes.

I agree with you; I can do some refactor, but then it will affect everyone...

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega commented:
> In my opinion this capability should be implemented in the base analysis/solver. I just cannot see how our abstract classes are general enough if they cannot include remeshing naturally.
> 
> However, I this way of extending the capabilities of the base classes (through a function wrapper) can work as a way to test them before they can be added to the base classes.

Agree, the @KratosMultiphysics/implementation-committee waits for the point of @adityaghantasala to reach a consensus

loumalouomega commented:
Ping @adityaghantasala

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

adityaghantasala commented on kratos/python_scripts/auxiliar_methods_adaptative_remeshing.py:
```diff
@@ -0,0 +1,161 @@
+                KratosMultiphysics.Logger.PrintInfo(analysis._GetSimulationName(), "Adaptative strategy not converged after ", non_linear_iteration, "iterations" )
+                break
+            else:
+                if with_contact:
```
probably this will be gone. But `with_contact` should not be in the core uitility.

loumalouomega replied:
True, I will think on an alternative

adityaghantasala commented on kratos/python_scripts/auxiliar_methods_adaptative_remeshing.py:
```diff
@@ -0,0 +1,161 @@
+    while analysis.KeepAdvancingSolutionLoop():
+        analysis.time = solver.AdvanceInTime(analysis.time)
+        non_linear_iteration = 1
+        while non_linear_iteration <= analysis.non_linear_iterations:
```
which non-linear iteration is this ? for the Newton-raphson  ?

loumalouomega replied:
Yes

adityaghantasala commented on kratos/python_scripts/auxiliar_methods_adaptative_remeshing.py:
```diff
@@ -0,0 +1,161 @@
+                non_linear_iteration += 1
+        analysis.OutputSolutionStep()
+
+def ReInitializeSolver(analysis):
```
Such a function can be useful in the base solver.

loumalouomega replied:
That would take long to be accepted

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

AlejandroCornejo commented:
The changes proposed by @adityaghantasala should be implemented before we approve this PR.

AlejandroCornejo reviewed (changes requested):
See last comments about the changes required by @adityaghantasala

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Moving ReInitializeSolver to analysis

loumalouomega committed:
Many refactor

loumalouomega committed:
Refactor

loumalouomega commented:
@adityaghantasala I addressed your comments

RiccardoRossi commented:
i think this is interesting, nevertheless i think i need some explanations. 
@loumalouomega let's organize a meeting so that you explain me a bit the design. Of course let's make it so that anyone interested can join.

loumalouomega commented:
> i think this is interesting, nevertheless i think i need some explanations.
> @loumalouomega let's organize a meeting so that you explain me a bit the design. Of course let's make it so that anyone interested can join.

OK

philbucher commented:
I would like to join the meeting

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega commented:
We still have pending the meeting about this

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega commented:
@KratosMultiphysics/technical-committee  just to remember that the @KratosMultiphysics/implementation-committee  reached a consensus long time ago

loumalouomega committed:
Update Structural script

loumalouomega committed:
Pythonic import (I)

loumalouomega committed:
Typo fix on mmg_process.py

loumalouomega committed:
Problems on the pythonic update (I)

loumalouomega committed:
Problems on the pythonic update (II)

loumalouomega committed:
Problems on the pythonic update (III)

loumalouomega committed:
[Core] Preventing potential error when not initialized on SImpleMortarMapper

INDEX_SET may not be initialized, this initializes on that case

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
[Core] Adding missing arrays to kratos_globals

Array4,Array6,Array9 were missing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
[Core] Extending python export of SetNonHistoricalVariableToZero to more array types

This extends for Array4, Array6, Array9

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
[MeshingApplication] MMGProcess restores metric after remesh

Basically that. Until now it was taking the interpolated metric, which does not necessary coincides with the correct new metric

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Update parameters

loumalouomega committed:
[MeshingApplication] Adding EnsureProperties method

This ensures that the properties of the elements and conditions are defined on the model part

loumalouomega committed:
Fix

loumalouomega committed:
Ensuring properties

loumalouomega committed:
Adding test

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Update method name

loumalouomega committed:
Using submodule

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Update script

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Update to recent changes

AlejandroCornejo reviewed (approved):
ok!

AlejandroCornejo reviewed (changes requested):
sorry I only saw the last commit, let's wait for the rest :P

loumalouomega commented:
> sorry I only saw the last commit, let's wait for the rest :P

I am fixing this branch right now, need to be up-to-date

loumalouomega committed:
Update settings

loumalouomega committed:
Merge master

loumalouomega committed:
[ContactStructuralMechanicsApplication] Replace error fro warning

This doesn't allow to remesh contact problems

RiccardoRossi commented on applications/FluidDynamicsApplication/python_scripts/fluid_dynamics_analysis.py:
```diff
@@ -8,6 +8,9 @@
 from KratosMultiphysics.FluidDynamicsApplication import python_solvers_wrapper_fluid
 
+# Import auxiliar methods
+from KratosMultiphysics.auxiliar_methods_adaptative_remeshing import AuxiliarMethodsAdaptiveRemeshing
```
thie should be approved by the flui dynamics team

@rubenzorrilla @jcotela you ok with this?

rubenzorrilla replied:
Before any specific approval I'd rather discuss this in the @KratosMultiphysics/technical-committee .

RiccardoRossi commented on applications/StructuralMechanicsApplication/python_scripts/adaptative_remeshing_structural_mechanics_implicit_dynamic_solver.py:
```diff
@@ -45,8 +35,7 @@ def __init__(self, model, custom_settings):
     def AddVariables(self):
         super(AdaptativeRemeshingImplicitMechanicalSolver, self).AddVariables()
-        if not missing_meshing_dependencies:
-            self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.NODAL_H)
```
was this approved by the StructuralMechanics team? @philbucher ?

loumalouomega replied:
It is an independent solver

philbucher replied:
to me ok, this is in the remeshing. If it needed there then ok

RiccardoRossi commented on applications/StructuralMechanicsApplication/python_scripts/adaptative_remeshing_structural_mechanics_analysis.py:
```diff
@@ -2,15 +2,19 @@
 from KratosMultiphysics.StructuralMechanicsApplication.structural_mechanics_analysis import StructuralMechanicsAnalysis as BaseClass
 
+# Import solver wrapper
+from KratosMultiphysics.StructuralMechanicsApplication import python_solvers_wrapper_adaptative_remeshing_structural
```
did structural mechanics approve?

loumalouomega replied:
@KratosMultiphysics/implementation-committee did it three months ago

philbucher replied:
I think this is ok since it is the remeshing-analysis

RiccardoRossi commented on kratos/python_scripts/analysis_stage.py:
```diff
@@ -191,6 +191,28 @@ def ChangeMaterialProperties(self):
         """this function is where the user could change material parameters as a part of the solution step """
         pass
 
+    def ReInitializeSolver(self):
```
this should definitely approved by @KratosMultiphysics/technical-committee

philbucher replied:
I am really scared of this one :)

loumalouomega replied:
There is no other option

RiccardoRossi replied:
I don't think this is correct. For example imagine that you arrive to a given level of deformation. If you reinitialize the elements that initial deformation would be considered as the reference.

RiccardoRossi reviewed (changes requested):
identifying discussion points

loumalouomega committed:
Update structural scripts

loumalouomega committed:
Update contact scripts

philbucher commented:
@KratosMultiphysics/technical-committee this is very intrusive, I think we should first discuss this globally before the individual apps approve

@loumalouomega is this tied to mmg or is it also possible to use a different tool for remeshing?

loumalouomega commented:
> @KratosMultiphysics/technical-committee this is very intrusive, I think we should first discuss this globally before the individual apps approve

This was discussed long time ago in the @KratosMultiphysics/implementation-committee, and we reached consensus. We were waiting for @KratosMultiphysics/technical-committee 

> @loumalouomega is this tied to mmg or is it also possible to use a different tool for remeshing?

It is generic, but maybe the criteria considered can be discussed. Right now the system is reinitialized if the modelpart has set the flag MODIFIED

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Some cleanings

loumalouomega committed:
Adding ClearDatabase

loumalouomega committed:
Using ClearDatabase

loumalouomega committed:
Better analysis

loumalouomega committed:
Cleaning

loumalouomega committed:
Document

loumalouomega committed:
Typo

loumalouomega committed:
Suggestions

loumalouomega committed:
More consistent with Initialize

loumalouomega committed:
Adding flag to avoid duplicated remesh

philbucher commented:
If this becomes a standard feature in the core, then I would propose to add a flag REMESHED

loumalouomega commented:
> If this becomes a standard feature in the core, then I would propose to add a flag REMESHED

You are opening the box of Pandora proposing a new flag

loumalouomega committed:
Update script

loumalouomega committed:
Ensuring properties

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Adding _AuxiliarCallsBeforeRemesh

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Preserving initial minimal and maximal size

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega commented:
Can we proceed with this?

loumalouomega commented:
@KratosMultiphysics/implementation-committee has agreed that we should proceed with the final discussion regarding this PR. The number of conflicts will increase more and more with the time

@KratosMultiphysics/technical-committee @philbucher 

BTW: All the cases in the Examples Repository work and only work with this branch

RiccardoRossi commented on kratos/python_scripts/analysis_stage.py:
```diff
@@ -193,6 +193,49 @@ def ChangeMaterialProperties(self):
         pass
 
+    def ClearDatabase(self):
+        """ This method clears the database in case it is necessary
```
what does it mean to "clear the database"?

loumalouomega replied:
By default does nothing, but in the contact I used in order to avoid conflicts (I remove the contact conditions and I create again later)

GuillermoCasas replied:
In the @KratosMultiphysics/implementation-committee we have looked and it does not appear that you actually make use of 'ClearDatabase' in your implementations. The name is too vague, inviting developers to make dangerous implementations. Could elaborate on what actually you think this function should be able to do? What database are you thinking about?

RiccardoRossi reviewed (commented):
i will comment post a detailed comment asap

RiccardoRossi commented:
I am still not convinced on this, particularly on the "ClearDatabase" and "Reinitialize"

As I understand what you need is a "ResetSolutionStep", to rewind the current step to its beginning so that boundary conditions will be reapplied. 

the overall flow would then be something like

~~~py
    def RunSolutionLoop(self):
        """This function executes the solution loop of the AnalysisStage
        It can be overridden by derived classes
        """
        while self.KeepAdvancingSolutionLoop():

            adaptivity_needed = True
            while(adaptivity_needed):
                  self.time = self._GetSolver().AdvanceInTime(self.time)
                  self.InitializeSolutionStep()
                  self._GetSolver().Predict()
                  is_converged = self._GetSolver().SolveSolutionStep()
                  self.__CheckIfSolveSolutionStepReturnsAValue(is_converged)

                  ##here check adaptivity
                  if(... adaptivity is needed .... )
                         eventually Remesh AND INITIALIZE NEWLY CREATED ELEMENTS
                         strategy.clear()
                         ResetSolutionStep  ... here i mean go back to the status at the very beginning of the time step
                         Adapt Time Step  ---> this i am adding ... if you do not get convergence you will also want to adapt the time step - for example this could be done by default if the max iterations are exceeded (something that already gives a lot of robustness)

                         adaptivity_needed = True
                   else:
                         adaptivity_needed = False

            self.FinalizeSolutionStep()
            self.OutputSolutionStep()
~~~

am i missing something in this picture?

loumalouomega commented:
@RiccardoRossi I cannot reply you directly because I don't know how do you create your comments. The ReInitialize is necessary because the elements and conditions are new, and therefore we need to initialize them (and not only the elements and conditions), many operations of initialization are done in the solvers and processes. If I do as you suggest it may not work in cases where the elements are not standard and require additional operations, doing it this way guarantees that it will work for any case

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

RiccardoRossi commented:
@loumalouomega if you see in my comment i write

"eventually Remesh AND INITIALIZE NEWLY CREATED ELEMENTS"

I dothink that if you remesh the only thing that should be programmatically reinitialized are the elements and conditions. 
As a minimum the "reinitialization" to be performed CAN (i would say SHOULD) be different from a complete reinitialization of the solvers. 

Can u make examples of which things need to be done in the Reinitialize concretely? (other than the elements and conditions...for which btw we need to find a way to prescribe initial deformations)

loumalouomega commented:
> @loumalouomega if you see in my comment i write
> 
> "eventually Remesh AND INITIALIZE NEWLY CREATED ELEMENTS"
> 
> I dothink that if you remesh the only thing that should be programmatically reinitialized are the elements and conditions.
> As a minimum the "reinitialization" to be performed CAN (i would say SHOULD) be different from a complete reinitialization of the solvers.
> 
> Can u make examples of which things need to be done in the Reinitialize concretely? (other than the elements and conditions...for which btw we need to find a way to prescribe initial deformations)

For example in the contact (but not only). I remove all the contact conditions because otherwise MMG makes strange things (@marcnunezc  started recently  #5605
about this problem), after remesh I recover the condition reinitializing the processes and solvers. Additionally many processes expect certain information that may not exist anymore, and therefore the only way to ensure it behaves as expected.

About the initial deformations, I will say it again, you need to use the processes that exist to transfer GP values, there is no other way.

Can you give a reason why reinitialize is not good option?

RiccardoRossi commented:
Reinitializing everything is the simplest solution, but i do not think it is the best. 

I am of the opinion that one should carefully reinitialize only what is strictly needed

For example if you use an acceleration strategy (to make an example in the context of FSI) you want to keep the history of what you have been doing in the last steps, reinitializing the solver you would lose that (every step would be like the first step).

Also the treatment of internal variables should be done differently when you restart a step and when
you want to Initialize the entire solver.

loumalouomega commented:
> Reinitializing everything is the simplest solution, but i do not think it is the best.
> 
> I am of the opinion that one should carefully reinitialize only what is strictly needed
> 
> For example if you use an acceleration strategy (to make an example in the context of FSI) you want to keep the history of what you have been doing in the last steps, reinitializing the solver you would lose that (every step would be like the first step).
> 
> Also the treatment of internal variables should be done differently when you restart a step and when
> you want to Initialize the entire solver.

OK, what if by default I reinitialize just the minimum required, but we can override those methods in derived classes in case it is necessary?

RiccardoRossi commented:
I think it is a good idea, nevertheless 
1 - think about the naming of that function, so fill the proposal i was posting with dummy functions before doing more work
2 - let's see if we have an agreement on the formal structure (which i understand would go in the AnalysisStage)

loumalouomega committed:
Adding entities utilities

loumalouomega committed:
Initialize only elements and conditions

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega commented:
@RiccardoRossi done, now it just initializes the mininum (I used the fluid as reference, and I added the minimal initializations until the 2D channel worked, BTW this is also the minimum required by the contact).

loumalouomega commented:
> @RiccardoRossi done, now it just initializes the mininum (I used the fluid as reference, and I added the minimal initializations until the 2D channel worked, BTW this is also the minimum required by the contact).

In fact the contact required less, as the system is rebuild each time step, but in the fluid as this is not the case by default I reset to clear the solver (as the size of the system has changed)

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
[MeshingApplication] Update test after the new pythonic structure

loumalouomega committed:
Refactor analysis

loumalouomega committed:
Remove wrapper

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Update stuff

loumalouomega committed:
Merge master

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega commented:
Ping

loumalouomega commented:
Ping @RiccardoRossi

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega commented:
Ping @RiccardoRossi

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Fix

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

adityaghantasala commented on kratos/utilities/entities_utilities.cpp:
```diff
@@ -0,0 +1,211 @@
+namespace EntitiesUtilities
+{
+
+void InitializeEntities(ModelPart& rModelPart)
```
Can't  you  all `InitializeElements` , `InitializeCondition` and `InitializeMasterSlaveConstraints` functions inside this function ? that way you can reduce reimplementing ?

loumalouomega replied:
It is done in a independent PR, I am waiting fro approval #6389

adityaghantasala replied:
Sorry then I did not see #6389 . That addresses my questions here !

GuillermoCasas commented on kratos/python_scripts/auxiliar_methods_adaptative_remeshing.py:
```diff
@@ -0,0 +1,151 @@
+                    root_model_part.Set(KratosMultiphysics.MODIFIED, False)
+                    self.analysis.is_printing_rank = False
+                computing_model_part.ProcessInfo.SetValue(KratosMultiphysics.NL_ITERATION_NUMBER, non_linear_iteration)
+                is_converged = convergence_criteria.PreCriteria(computing_model_part, builder_and_solver.GetDofSet(), mechanical_solution_strategy.GetSystemMatrix(), mechanical_solution_strategy.GetSolutionVector(), mechanical_solution_strategy.GetSystemVector())
```
Should you be allowed to change the convergence criteria from here?

loumalouomega replied:
This is a call for the PreCriteria, not a modification

GuillermoCasas commented on kratos/python_scripts/auxiliar_methods_adaptative_remeshing.py:
```diff
@@ -0,0 +1,151 @@
+        mechanical_solution_strategy = solver.get_mechanical_solution_strategy()
+
+        while self.analysis.KeepAdvancingSolutionLoop():
+            self.analysis.time = solver.AdvanceInTime(self.analysis.time)
```
This is a full-fledged analysis stage... Could you not include the outermost functions in your own analysis stage and break down the most reusable parts into processes for everyone to use? (from @KratosMultiphysics/implementation-committee )

loumalouomega replied:
I do not understand

loumalouomega committed:
Merge master

loumalouomega committed:
Duplicated

loumalouomega committed:
Update name

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

GuillermoCasas commented on applications/ContactStructuralMechanicsApplication/python_scripts/auxiliar_methods_contact_adaptative_remeshing.py:
```diff
@@ -0,0 +1,162 @@
+            # We reinitialize if remeshed previously
+            if root_model_part.Is(KM.MODIFIED):
+                self.analysis.ReInitializeSolver()
+            self.ExecuteBeforeFinalizeSolutionStep()
```
Should this not be 'ExecuteBefore**Initialize**SolutionStep()'?

loumalouomega replied:
True; in fact the method is duplicated with the base class, I remove it

loumalouomega replied:
And I added the missing method

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Adding ExecuteInitializeSolutionStep

loumalouomega committed:
Update

loumalouomega committed:
Remove duplicated

loumalouomega committed:
Revert difference

loumalouomega committed:
Upload files

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Update utilities

loumalouomega committed:
Update search

GuillermoCasas commented on kratos/python_scripts/auxiliar_methods_adaptative_remeshing.py:
```diff
@@ -0,0 +1,161 @@
+        computing_model_part = self.analysis._GetSolver().GetComputingModelPart()
+        root_model_part = computing_model_part.GetRootModelPart()
+
+        while self.analysis.KeepAdvancingSolutionLoop():
```
The original idea was that 'RunSolutionLoop' was not to be reimplemented. This workaround that does precisely that. Should we not be improving the design of the original function instead of using it like this?

loumalouomega replied:
I will answer you in the second question

GuillermoCasas commented on kratos/python_scripts/auxiliar_methods_adaptative_remeshing.py:
```diff
@@ -0,0 +1,161 @@
+                    non_linear_iteration += 1
+            self.analysis.OutputSolutionStep()
+
+    def ExecuteInitializeSolutionStep(self):
```
Should this function not be called 'Execute**Before**InitializeSolutionStep'? Could it not be added to the base analysis_stage if needed? Idem with ExecuteBeforeFinalizeSolutionStep.

GuillermoCasas commented on kratos/python_scripts/auxiliar_methods_adaptative_remeshing.py:
```diff
@@ -0,0 +1,161 @@
+            self.analysis.time = self.analysis._GetSolver().AdvanceInTime(self.analysis.time)
+            # We reinitialize if remeshed previously
+            if root_model_part.Is(KratosMultiphysics.MODIFIED):
+                self.analysis.ReInitializeSolver()
```
Can you not make this call (and all the subsequent ones) inside a reimplementation of <strike>ExecuteInitializeSolutionStep</strike>InitializeSolutionStep? in a remeshing analysis that derives from the base one?

loumalouomega replied:
OK, maybe you are right

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Correct name

loumalouomega committed:
Unify

loumalouomega committed:
Move method

loumalouomega commented:
@GuillermoCasas I applied your suggestions

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Revert modification

loumalouomega committed:
Correct script

loumalouomega committed:
Missing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Fix to be compatibl with CoSIm app

GuillermoCasas commented on kratos/python_scripts/analysis_stage.py:
```diff
@@ -202,6 +219,74 @@ def ChangeMaterialProperties(self):
+        # Clear solver
+        solver.Clear()
+
+        # Some initilizations
```
All these different options should be implemented in derived analyses stages, not the base one.

GuillermoCasas commented on kratos/python_scripts/analysis_stage.py:
```diff
@@ -60,12 +60,29 @@ def RunSolutionLoop(self):
+        if hasattr(self, 'model'):
+            model_part_names = self.model.GetModelPartNames()
+            for model_part_name in model_part_names:
+                if self.model.GetModelPart(model_part_name).Is(KratosMultiphysics.MODIFIED):
```
Should the tag name not be 'MODIFIABLE'?. Otherwise the natural interpretation is of something that may or may not change at each time step.

loumalouomega replied:
The proper thing woul be to create a new flag, but requires a whole discussion

GuillermoCasas commented on kratos/python_scripts/analysis_stage.py:
```diff
@@ -60,12 +60,29 @@ def RunSolutionLoop(self):
         while self.KeepAdvancingSolutionLoop():
-            self.time = self._GetSolver().AdvanceInTime(self.time)
+            self.time = solver.AdvanceInTime(self.time)
+            # We reinitialize if remeshed previously
```
the tag 'modified_model' is not changing its value during the simulation...

loumalouomega replied:
True

loumalouomega replied:
I modified, sorry, I run a test with just one step...

loumalouomega committed:
Defining __CheckIfModelIsModified

adityaghantasala commented on applications/ContactStructuralMechanicsApplication/python_scripts/adaptative_remeshing_contact_structural_mechanics_analysis.py:
```diff
@@ -22,158 +27,133 @@ def __init__(self, model, project_parameters):
+            "analysis_type"    : "linear",
             "contact_settings" :
             {
                 "fancy_convergence_criterion" : false
```
I will suggest to remove the prefix `fancy_`

loumalouomega replied:
It is alrady in my app, not related with this PR (is not in red or green)

philbucher replied:
we have the `KRATOS_COLORED_OUTPUT` definition in the meantime which should be used for this

adityaghantasala replied:
@loumalouomega its about usability, it not common to have `fancy_` as a prefix also it is not explaining anything. my be with the suggestion of philipp you can altogether remove this setting.
of course you can choose any name you want for the key as I put it there, its a suggestion !

loumalouomega replied:
> 
> 
> we have the `KRATOS_COLORED_OUTPUT` definition in the meantime which should be used for this

This is not related with this PR in any case, and is not the same as the colours, is the structure I use table like

adityaghantasala commented:
On behalf of @KratosMultiphysics/implementation-committee 

We require the following changes : 
- We see that the `RunSolutionLoop` function is used improperly. If something is lacking in the base analysis stage and needed to be added to be more flexible, it should be pointed out and added in a different PR. 

- The adaptivity of the mesh should be centralized so that all the required operations are done at a single point. 

- It is already mentioned in the comments above,  `ClearDataBase` function should be redefined to be a more specific version (that is if it is really necessary). For example making it a private function in the base case.

Here are some suggestions to improve the design : 

- The non-linear iteration loop in `SPRAdaptativeRemeshingRunSolutionLoop` should be moved into the solver (in python or in C++).

- A possible redesign can be to move all the necessary steps of remeshing into `InitializeSolutionStep` of the python solver. So that there is only one point where remeshing is done. 

- Please have a look at the comment
https://github.com/KratosMultiphysics/Kratos/pull/4558#discussion_r400200246
May be some functions can be added to base python solver to make this possible in a nice way.

loumalouomega committed:
Merge master

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Minor

loumalouomega commented:
@adityaghantasala after reading the comments of the @KratosMultiphysics/implementation-committee I have several comments. But basically everything can be summarized as the follows:

You cannot move the remeshing to the solver because you require to re-run all processes, and the solver does not know the processes, so it can be done ONLY in the analysis stage. 

For the other comments I do not know how to adapt it, and basically is the same reason, I require the processes to be reevaluated again

loumalouomega committed:
Refactor analysis

loumalouomega committed:
Update analysis

loumalouomega committed:
Simplify

loumalouomega committed:
Minor

loumalouomega committed:
Recursive error

loumalouomega committed:
Update clean up

loumalouomega committed:
[MeshingApplication] Some improvements/additional cheks in MMGProcess

Just that

loumalouomega committed:
Clean up

loumalouomega committed:
Adding CleanContactModelParts

loumalouomega committed:
Update

loumalouomega committed:
More update

loumalouomega commented:
@KratosMultiphysics/implementation-committee I refactored and cleaned up this the most I could. I created #6764 and #6762 in order to reduce the complexity of this PR

adityaghantasala commented:
> For the other comments I do not know how to adapt it, and basically is the same reason, I require the processes to be reevaluated again

You mean you want to re-evaluate yourself again no ?

loumalouomega commented:
> > For the other comments I do not know how to adapt it, and basically is the same reason, I require the processes to be reevaluated again
> 
> You mean you want to re-evaluate yourself again no ?

I mean that it must be at the analysis in order to run the processes (otherwise BC are lost for example)

adityaghantasala commented:
> The non-linear iteration loop in SPRAdaptativeRemeshingRunSolutionLoop should be moved into the solver (in python or in C++).

Did you explore the possibility of moving that function to C++ ?

loumalouomega commented:
> > The non-linear iteration loop in SPRAdaptativeRemeshingRunSolutionLoop should be moved into the solver (in python or in C++).
> 
> Did you explore the possibility of moving that function to C++ ?

I need the processes from the analysls

adityaghantasala commented:
> > > For the other comments I do not know how to adapt it, and basically is the same reason, I require the processes to be reevaluated again
> > 
> > 
> > You mean you want to re-evaluate yourself again no ?
> 
> I mean that it must be at the analysis in order to run the processes (otherwise BC are lost for 

Understood .. sorry I miss read your comment. 

Secondly, is this where you say processes are needed 
https://github.com/KratosMultiphysics/Kratos/blob/f399f90c462dfddbae536e419b984f2921b15a9d/kratos/python_scripts/auxiliar_methods_adaptative_remeshing.py#L105

isn't this because that there is different mesh refinement/adaptivity process. If this is some how integrated into the new C++ strategy class you will not need it here !

loumalouomega commented:
> > > > For the other comments I do not know how to adapt it, and basically is the same reason, I require the processes to be reevaluated again
> > > 
> > > 
> > > You mean you want to re-evaluate yourself again no ?
> > 
> > 
> > I mean that it must be at the analysis in order to run the processes (otherwise BC are lost for
> 
> Understood .. sorry I miss read your comment.
> 
> Secondly, is this where you say processes are needed
> 
> https://github.com/KratosMultiphysics/Kratos/blob/f399f90c462dfddbae536e419b984f2921b15a9d/kratos/python_scripts/auxiliar_methods_adaptative_remeshing.py#L105
> 
> isn't this because that there is different mesh refinement/adaptivity process. If this is some how integrated into the new C++ strategy class you will not need it here !

For the SPR i cannot not do it in a C++ way, because the processes are available only in the analysis in python, and for the SPR remeshing the remesh is done after each NL iteration, not at the end/begining of the solution step. It cannot be done other way

loumalouomega committed:
Fix (I)

loumalouomega committed:
Fix (II)

loumalouomega committed:
Adding additional check

loumalouomega committed:
Clean up

loumalouomega committed:
Additional checks

philbucher commented on kratos/python_scripts/analysis_stage.py:
```diff
@@ -124,12 +124,22 @@ def InitializeSolutionStep(self):
+        # We reinitialize if remeshed previously
+        if self._CheckIfModelIsModified():
+            self.ClearDatabase()
+            self.ReInitializeSolver()
```
I am quite convinced this cannot be in the baseclass

loumalouomega replied:
Well, I am quite sure we agreed this in order to be used with the corresponding NR for an adaptive strategy

philbucher commented on applications/FluidDynamicsApplication/python_scripts/fluid_dynamics_analysis.py:
```diff
@@ -39,7 +39,7 @@ def _CreateProcesses(self, parameter_name, initialization_order):
 
         if parameter_name == "processes":
-            processes_block_names = ["gravity", "initial_conditions_process_list", "boundary_conditions_process_list", "auxiliar_process_list"]
+            processes_block_names = ["gravity", "initial_conditions_process_list", "boundary_conditions_process_list", "auxiliar_process_list","mesh_adaptivity_processes"]
```
this can be removed, in fact is deprecated and throws an error now

loumalouomega replied:
OK

philbucher commented on applications/FluidDynamicsApplication/python_scripts/fluid_dynamics_analysis.py:
```diff
@@ -71,7 +71,8 @@ def _GetOrderOfProcessesInitialization(self):
                 "boundary_conditions_process_list",
-                "auxiliar_process_list"]
+                "auxiliar_process_list",
+                "mesh_adaptivity_processes"]
```
I would really really not add anything unless absolutely necessary
I think the order of when those processes are done doesn't matter

loumalouomega replied:
I think is adequate to have a category just for mesh adaptivity

philbucher replied:
read the doc in the baseclass:
~~~
This function can be overridden in derived classes if the order of initialization for the processes matters
~~~

this should NOT be necessary!!!

The names of the blocks don't matter otherwise

loumalouomega replied:
The problem is then the order?, in any case remeshing should be the last thing to do

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/adaptative_remeshing_structural_mechanics_analysis.py:
```diff
@@ -2,15 +2,19 @@
 
```
I am thinking that in order to improve the organization all the `adaptive_remeshing` things can go into a `adaptive_remeshing` folder
same I wanted to do for a while now with the Solvers, but haven't gotten to do it yet

loumalouomega replied:
OK

loumalouomega replied:
Done

philbucher commented on kratos/python_scripts/auxiliar_methods_adaptative_remeshing.py:
```diff
@@ -0,0 +1,112 @@
+        computing_model_part = solver.GetComputingModelPart()
+        convergence_criteria = solver.get_convergence_criterion()
+        builder_and_solver = solver.get_builder_and_solver()
+        mechanical_solution_strategy = solver.get_mechanical_solution_strategy()
```
This is an implementational detail!

loumalouomega replied:
¿?, I don't understand the issue

philbucher replied:
this function is not in the Interface of `PythonSolver`

loumalouomega replied:
You mean we removed GetComputingModelPart?

loumalouomega replied:
GetComputingModelPart exists in the base solver, line 154

philbucher replied:
but e.g. `get_mechanical_solution_strategy` is not

loumalouomega replied:
OK, I will move this to structural auxiliar utilities

philbucher reviewed (changes requested):
overall this seems to have converged more now. I still have concerns regarding the implementation:
- The design seems to be very specific to the currently available options (basically MMG) and it does not seem to be flexible/extensible to other procedures (e.g. you are using implementational details e.g. from the structural solver)
=> maybe the `Modeler` concept @tteschemacher could help here? IMO this would fit with the design
- Using implementational details is ok if there is no other way for now, but then this can only live in the corresponding application until a better solution is found.
- I REALLY DON'T like the changes in the `AnalysisStage`, those are way too specific. Remember that the `AnalysisStage` is the baseclass for ALL analysis-stages in Kratos, also DEM, MeshMoving, CoSim, FSI => in many of them remeshing is not required/applicable.
Furthermore the `AnalysisStage` is more an interface (except the handling of the Processes), the actual implementation should not be done here
I am perfectly fine with having the (I understand necessary/required) additions in a separate file or derived class and then adding e.g. if necessary a new Interface function to the `AnalysisStage`
- In the beginning you wrote that this also works for fluid, but I see no modification for it. What is the status there?

Blocking until the concerns are addressed

@KratosMultiphysics/technical-committee I think we should discuss this in one of the next meetings

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Suggestion

loumalouomega commented:
>     * The design seems to be very specific to the currently available options (basically MMG) and it does not seem to be flexible/extensible to other procedures (e.g. you are using implementational details e.g. from the structural solver)
>       => maybe the `Modeler` concept @tteschemacher could help here? IMO this would fit with the design

What do you suggest instead'?

> 
>     * Using implementational details is ok if there is no other way for now, but then this can only live in the corresponding application until a better solution is found.

OK

> 
>     * I REALLY DON'T like the changes in the `AnalysisStage`, those are way too specific. Remember that the `AnalysisStage` is the baseclass for ALL analysis-stages in Kratos, also DEM, MeshMoving, CoSim, FSI => in many of them remeshing is not required/applicable.

But by default preserves current behavior. What do you suggest?, an intermediate analysis where fluid and structural derive?

>       Furthermore the `AnalysisStage` is more an interface (except the handling of the Processes), the actual implementation should not be done here

The problem is that I require the processes in order to reset BC and other settings. Any other suggestion is welcome

>       I am perfectly fine with having the (I understand necessary/required) additions in a separate file or derived class and then adding e.g. if necessary a new Interface function to the `AnalysisStage`
> 

OK

>     * In the beginning you wrote that this also works for fluid, but I see no modification for it. What is the status there?
> 

It works, you can try the examples in the Example repository. The fluid requires less modifications because doe snot have SPR process. With the base analysis changes is enough

> 
> Blocking until the concerns are addressed

OK

> 
> @KratosMultiphysics/technical-committee I think we should discuss this in one of the next meetings

OK

loumalouomega committed:
Moving scripts to a folder (I)

loumalouomega committed:
Moving scripts to a folder (II)

loumalouomega committed:
Update

loumalouomega committed:
Moving SPRAdaptativeRemeshingRunSolutionLoop

philbucher commented:
I think having an intermediate class for these functionalities would be a good solution until maybe in the future it can be done better with the a `Modeler`

loumalouomega commented:
> I think having an intermediate class for these functionalities would be a good solution until maybe in the future it can be done better with the a `Modeler`

Can you detail the class structure?

loumalouomega committed:
Merge master

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Codacy suggestion (I)

loumalouomega committed:
Codacy suggestion (II)

loumalouomega committed:
Codacy suggestion (III)

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

GuillermoCasas commented:
It would be very nice if you broke down this PR into two: one for the proposed changes to the core and another for the changes to StrucuralMechanics. I think it will facilitate getting through, since the shorter core changes will attract more attention. Once these are passed, the rest will be easier. Just a suggestion.

AlejandroCornejo commented:
On behalf of @KratosMultiphysics/implementation-committee We agree with @GuillermoCasas, this PR should be divided in order to separate more clearly the changes in the core and SMA. Thank you @loumalouomega

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega commented:
@KratosMultiphysics/dem tests are still failing in windows https://github.com/KratosMultiphysics/Kratos/pull/4558/checks?check_run_id=614373750

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Refactor

loumalouomega committed:
Merge master

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
More modular design as suggested by @GuillermoCasas

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

loumalouomega committed:
Update

loumalouomega committed:
[MeshingApplication] Adding clear which avoids errors in dofs

This avoids some potential errors

loumalouomega committed:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

Commit:
Revert "Revert "add function AddColors to the communicator""

This reverts commit 28ea97aaa1488fc95f038b7deeae947dc2bd9ed7.

Commit:
Merge master

Commit:
Update

Commit:
Merge branch 'master' into several_applications/missing_scripts_adaptative_remeshing

Commit:
Merge master

Commit:
Fix

Commit:
Clean up
